### DISCUSSION TITLE
DirectX and COM interoperability for .NET

### DISCUSSION THREAD

**Discussion by atcarter714** [microsoft/CsWin32#705] [General]
So, I found my way here after a lot of research, landing on the WinMD repo and being directed to @AArnott by Mr. Kerr. I've managed to get CsWin32 to work for me for a simple test of DXGI and enumerating my adapters and outputs. This is lovely! I support this effort/project to the fullest because the possibilities here are truly amazing. It was like a Hollywood villain moment when I saw all those Win32 and COM APIs in C# IntelliSense! The power! 😄 

So I would like to know, first of all: What can I do to help, support and contribute to this? I think bringing full DirectX to .NET is an important and noble effort, and this project is a godsend for that sort of thing. I'm more than happy to convert existing DirectX sample code over to C# with this, for example, that shouldn't be super hard. I can also possibly contribute to the code generation system as well as supplying feedback and testing things. Basically, whatever needs doing I'm willing to take a shot at it. Really interested in learning more about this and doing whatever I can to help things along and make improvements.

I'm sure to have a ton of questions. One I've got now is what is the most elegant way to get the RRID / GUID / UUID or whatever name you prefer of these COM interfaces. I couldn't figure out an obvious way other than clicking on _Go to Definition_ and copying the Guid attribute string and using Guid.Parse. Ideally, this is something I think could be handled in a cleaner and more ".NET-like" way like a generic/Type argument (like with a _where T: <ICOMBaseTypeName>_ constraint) and it would infer the GUID from the attribute. Barring that, what would be a nice way to get it in my own code? Should I use reflection at startup to find them all or something? Don't really care for that idea but it would work. 🤷 

Anyhow, I can't wait to have some fun with this converting some C++ samples to C# and building a game/3D framework and would love to assist! ❤️

**Comment by atcarter714** [comment]
Also, I expect HRESULT as a return type for many of these DirectX functions and COM methods but many of them seem to be changed to void. I made EnumAdapters fail and it threw an exception, which made me feel a little better about it. Is that the ideal way to handle that just use the try/catch mechanism instead of capturing HRESULTS like in C++ and checking return codes?

**Reply by atcarter714** [reply]
OK, I see the ComException type has way more info in it than a simple HRESULT including documentation help links and a message. That's actually ... beautiful 😍

**Comment by AArnott** [comment]
I'm gratified that you're pleased with CsWin32 so far.

> what is the most elegant way to get the RRID / GUID / UUID or whatever name you prefer of these COM interfaces.

The ".NET way" is like this: `typeof(ISomeInterface).GUID`.

> I expect HRESULT as a return type for many of these DirectX functions and COM methods but many of them seem to be changed to void.

That's another .NET convention. [Learn more about PreserveSig](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.dllimportattribute.preservesig). CsWin32 will default to using PreserveSig (which preserves the HRESULT return type) only for those members that the metadata describes as needing it. In the metadata, we annotate PreserveSig on methods that return non-S_OK success return codes (e.g. S_FALSE), which are otherwise impossible to detect in .NET. We also annotate APIs that *commonly* fail (e.g. when reaching the end of an enumeration). That leaves the exception-throwing behavior to those methods that one would not typically expect to fail.
But if you like PreserveSig and want to see it more, you can add a property like this to your NativeMethods.json file (see the readme):
```json
  "comInterop": {
    "preserveSigMethods": [
      "IEnumString",
      "ISpellChecker"
    ]
  }
```

> the ComException type has way more info in it than a simple HRESULT

There isn't anything there that your program couldn't obtain after detecting a failing HRESULT, but yes, it's nice when it comes for free.

> What can I do to help, support and contribute to this?

Well, I love to hear that! Please pick an issue and dig in. You might start with the [help wanted issues](https://github.com/microsoft/CsWin32/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22), but you don't have to restrict yourself to that issue label. I recommend of anything more than a trivial change, comment on the issue that you'd like to take it, and how you intend to resolve it (e.g. describe the design or changes you'll make) and wait for my confirmation before proceeding to reduce risk of wasted effort if someone else is already working on it or the approach isn't the way we'll want to accept a PR for it later.

I'm looking forward to hearing more from you!

**Reply by atcarter714** [reply]
Wow, I feel silly now knowing it's as simple as typeof(T).GUID, I thought it was gonna be a bit more complicated with COM involved 😅 

So, I guess what I want to do now is get my hands dirty using this awesome tool and learn all about how it works and to make some nice things with it. I found your YouTube video from about a year ago and gained a little additional enlightenment from it. And yeah, I just saw the thing about PreserveSig which I think will be great for situations like IDXGIFactoryX.EnumAdaptersX where I expect it to fail at least 1 in every 3 or 4 calls on average and return a DXGI_ERROR_NOT_FOUND when it goes past the index of the last adapter on the system. I'd like to figure out what other customization I can do to the generated code and figure out where I should draw the line with certain things.

One thing I'm doing is working on a neat, tidy .NET library for DirectX and then to implement some DirectX samples in C#. I think that I can even bolster some support in game dev communities for this project as well as my DirectX-specific efforts. And I think it'd be great to put out an example of DirectX12 running in a WinUI 3 window, maybe even implement a little 3D panel control you can just stick in your WinUI apps with a XAML snippet, configure its properties and then get some update/draw calls pumping ... so, theoretically, someone can drop DirectX functionality right into a WinUI 3 app with little effort other than worrying about what to draw and how to draw it! And this sort of experimental stuff should give me plenty of insight on the usage of CsWin32, identify bugs and problems and highlight opportunities to improve things. I'm going to explore the issues on the repo too and figure out where I can help improve things! I'll also note any DirectX interop problems I come across and we can work them out.

So, right now, I'm tinkering with creating a pretty C# assembly "DXSharp" where I'm wrapping stuff like DXGICreateFactoryX with a prettier CreateFactoryX<T> where T: IDXGIFactory method. That raises a few questions for me, like do we want CsWin32 to provide those more elegant wrappers or is that the responsibility of a developer (me, in this case).

What I'm after is a nice, pretty layer to expose DirectX functionality to .NET code in a manner that's intuitive to .NET devs and inline with C# conventions ... cutting out pointer usage that we can live without, automatically obtaining GUIDs from generic type arguments with constraints to the base type, neat namespaces and encapsulation and of course good names for types and methods. The idea is to keep it true to the "essence" of the native DirectX libraries but present it all in a neat, clean and .NET dev-friendly way sort of like SlimDX and SharpDX. And it presents some interesting questions about this from a user perspective ...

First concern I have is what sort of approach I should use to wrapping these native COM interfaces for this sort of library and still expose a way the consumer can grab a pointer or reference to it if they want to do their own low-level stuff with it. It kind of makes sense to a C++ dev why DXGI stuff lives in several header files but from a C# perspective its kind of nasty to have to have 4 or 5 namespace using directives to get all the DXGI stuff at your fingertips, haha. That's a good example of something I want to make a prettier wrapper for and keep in a DXSharp.DXGI namespace with nice outward-facing classes, structs, enums, etc. I'm just not quite sure how I should approach that yet. And I suppose I need some more reference on how to configure things with NativeMethods.txt/json files so I can tweak my signatures to perfection. I've picked up on a few things from the repo instructions and what I've seen/read around here but I am sure I only know a small part of the story as of yet. I've managed to make some pretty wrapper methods with generic syntax already though for certain things ... but I'm not 100% sure how I should go about some of these things.

**Reply by atcarter714** [reply]
By the way, how do I force a COM interface to be defined _partial_ so I can add overloads manually? And is there more comprehensive documentation on NativeMethods.json configuration to customize my code generation?

**EDIT:** Think I'm gonna just try out a few approaches and see what works out the best for this stuff. Also wondering if there's a way you can just dump all the generated code to .cs files somewhere (or is that a feature we could add?) if you wanted to take over it and handle-edit it into something? Right now, I'm trying out the approach of using containment of COM interfaces inside my own objects that implement IDisposable and can reveal a pointer or reference to the real COM object, so the user can _not_ worry about COM unless they want to do something special on their own.

And excuse my wall o' text posts, I'm pretty excited and hyped up about this and showing the game dev community some working examples of DirectX in a modern .NET 6 and 7 assembly :D

**Reply by AArnott** [reply]
> PreserveSig which I think will be great for situations like IDXGIFactoryX.EnumAdaptersX

When you think PreserveSig would make a good default for *every* user of an API, please file a bug in the win32metadata repo for consideration.

> do we want CsWin32 to provide those more elegant wrappers or is that the responsibility of a developer (me, in this case).

CsWin32 will stick very closely to the metadata. We do have limited functionality to go beyond that (e.g. friendly overloads) and a *few* places where we recognize specific types and emit specially-authored code. But writing "an elegant wrapper" is decidedly outside the scope of CsWin32. So please feel free to write your DxSharp library (which sounds great, BTW) without fear that CsWin32 will erode your turf significantly.

> what sort of approach I should use to wrapping these native COM interfaces for this sort of library and still expose a way the consumer can grab a pointer or reference to it if they want to do their own low-level stuff with it

That touches on the question of whether you should turn on _public_ CsWin32 API generation vs. leave it at its default `internal` generation. Do you want your wrapper assembly to expose the underlying APIs as well? Typically, we recommend the `internal` default. But that can be cumbersome because then your own API cannot 'leak' even simple underlying types like enums -- you'de have to define them all yourself. But making _all_ of them public not only means your library exposes two API sets, but also makes it potentially more difficult for your library users when they reference *two* libraries that both contain embedded CsWin32-generated APIs. Those users will have to set up C# assembly ref aliasing and have disambiguation hell to deal with. This is why we generally discourage emitting public APIs. What you can do though is copy a subset of the emitted APIs into your own library, and change them to be public. CsWin32 will suppress generation of any type that your project already defines. In this way, you can control the public APIs published by your library, and still get the benefit of CsWin32 for internal APIs, as well as the origin of your public APIs.

> how do I force a COM interface to be defined partial so I can add overloads manually?

You can't add methods to a COM interface. That would break the vtbl. But you can always add _extension_ methods, and you don't need a partial modifier on the interface for that. See also #324.

> And is there more comprehensive documentation on NativeMethods.json configuration to customize my code generation?

It's in the json schema. So long as your .json file references that as described in the readme, and you're using a decent JSON editor (VS Code and VS for Windows qualify) that reads the schema, you'll get completions and doc tips to help you know what's available.

> wondering if there's a way you can just dump all the generated code to .cs files somewhere

As a C# source generator, it doesn't do this directly. But you can examine all the generated code using Go To Definition, or expand Solution Explorer's Dependencies node to discover all the generated code:
![image](https://user-images.githubusercontent.com/3548/193909499-dcc42fab-cd56-4b26-ad71-db679f4add63.png)
If you really want to copy and paste all the code, what I would do is first set the option in NativeMethods.json to emit all code into one file. Then I'd open that one file and copy paste it all out.

**Reply by atcarter714** [reply]
I see, it will still be quite a bit of effort but CsWin32 is saving a tremendous amount of time here to make a DXSharp library. I just discovered that I can define my own definition and suppress the code generation and I did that with HRESULT, allowing me to customize some things and make it a public def. You're absolutely right, I don't want to dump the whole interop library into a public accessibility level because users will be confused and overwhelmed. It's a little tricky just doing it myself, but I'm making it less ambiguous and a little prettier by dropping the really verbose C++ namings like "IDXGIFactory7" and going to "IFactory7" ... it's already in the "DXSharp.DXGI" namespace so the elaborate name is no longer necessary or desirable in C#, and we can write some more concise but still expressive C# code this way! It also shouldn't be hard for C++ programmers to understand, I don't think, but open to opinions on that purely as a stylistic concern.

This is a really fun way to help get my bearings with CsWin32 and find ways to improve it though. I'll definitely have some suggestions for things as I work my way through DirectX.

**Reply by atcarter714** [reply]
By the way, yeah, I think the series of IDXGIFactoryX::EnumAdaptersX should probably be PreserveSig by default, since we're _anticipating_ a DXGI_ERROR_NOT_FOUND response (that's how it tell us we've run out of installed adapters). Same thing would apply to IDXGIAdapterX::EnumOutputsX and similar methods. Throwing an exception does work, but it's more desirable to return HRESULT, I think, and let the programmer write a bit cleaner logic. They can check it for DXGI_ERROR_NOT_FOUND vs another error code entirely and then decide if an exception really needs to be thrown.

**Reply by AArnott** [reply]
That's great feedback. Can you file an issue on win32metadata for that enum API preservesig recommendation?

**Reply by atcarter714** [reply]
https://github.com/microsoft/win32metadata/issues/1295#issue-1397082474

Hopefully I did that correctly. If not, forgive me and show me the way 😄

**Reply by atcarter714** [reply]
Thought of some other questions I had ...

What is the difference between this and something like CsWinRT projections? Does it use P/Invoke under the hood as well? Or is there a fundamental difference in how those projections work and any major performance differences? It's my understanding interop in .NET 6 and 7 is much more efficient/fast and I haven't quite wrapped my head around what's changed (like if the "thunk" has been reduced) and the WinRT stuff is all brand new to me ...

I'm also curious about the new delegate pointer feature and how it compares to doing a P/Invoke call in terms of performance. And how that might also compare to Marshal.GetDelegateForFunctionPointer to wrap a function pointer in a managed delegate (something I haven't had a reason to do in a good while). Just curious about these things since performance is going to come up anytime DirectX is involved and I'm not quite sure how these things compare now in the modern .NET versions.

**Reply by AArnott** [reply]
> Hopefully I did that correctly. If not, forgive me and show me the way

Correct, but a bit long-winded IMO. Your bug may get action more quickly if you make it a very quick read. For example:

> The following APIs return failing HRESULTs in primary scenarios and should therefore have PreserveSig=true set on them:
> * IDXGIFactory1::EnumAdapters1
> * IDXGIFactory2::EnumAdapters2

That's all you (or they) need. Listing out each one is probably best (unless it's an entire interface) since the changes to the repo are applied to each one and you don't want them to miss anything. Your (great) justification paragraphs *shouldn't* be needed at this point since everyone working in that repo are familiar with the problems, and just need users familiar with specific APIs to point out where the change needs to be made.

**Reply by AArnott** [reply]
> What is the difference between this and something like CsWinRT projections?

These two do not compete. They produce projections from very different metadata. It so happens that they also function and produce very differently, as described below:

Topic | CsWinRT | CsWin32
--|--|--
APIs | WinRT | Win32
When run | When the WinRT metadata is built | With the customer project
Form of projection | A shipping .NET assembly | C# source code that is compiled directly into the project
Runtime dependency? | Yes | No
Size of projection | Large (covers the entire API) | Small (covers exactly what is asked for)

**Reply by AArnott** [reply]
> I'm also curious about the new delegate pointer feature and how it compares to doing a P/Invoke call in terms of performance.

I don't expect there to be any perf difference, *assuming no marshaling is required for the p/invoke*. I *think* a function pointer doesn't support marshaling at all, so a fair comparison in perf would require that the p/invoke method would not do marshaling either. That's simple enough. For example, the signature should take `char*` instead of `string`. 

Delegates are slower than both of these. They add reflection on top of whatever you could do directly with function pointers or p/invoke.

CsWin32 also generates COM interfaces. This is where things get potentially different between .NET Framework and newer .NET 6+ runtime versions. There is a thunk between managed and native code when COM gets involved (even if no data marshaling is otherwise required). These come in the form or RCW and CCW. .NET Framework and .NET produce these on-demand, and that incurs a (small) up front perf hit that could be avoided if the assembly carried these thunks around in code already. In future .NET runtime versions where trimming and AOT are used, there isn't enough functionality left in the runtime (or metadata in the assembly?) to produce CCW/RCW at runtime, so they *must* be included in binary form in the dll. The new ComWrappers library helps to do this for such projects. 
CsWin32 does not help to make ComWrappers to work other than generate the interface and vtbl. We're looking at generating more code to help with ComWrappers and trimmable scenarios in the future.

**Reply by AArnott** [reply]
You're asking great questions here, but I think it would make it easier to track Q&A and help others if you start new Discussions for what could be argued are new top-level questions, even if your driving cause is all DirectX stuff. :)

**Reply by atcarter714** [reply]
Gotcha! Will work on organizing thoughts and questions a bit better, haha, I'm prone to drop text walls when I get started on something interesting. 😄

Really appreciate all the great answers to questions! I'll probably make a thread about interop performance concerns sometime soon when I get a little further along. This has been going very well so far though!

**Reply by atcarter714** [reply]
I was wondering who the issue belongs to when I find some kind of issue with generation and how to determine if it belongs to the Win32 metadata project or CsWin32 ... for example, I found some types getting generated as uint instead of proper enum types, and not sure who that belongs to or how to decide where to draw the line on the issues.

Also, what about auto-documentation issues? I'm sure that belongs here?

**Reply by atcarter714** [reply]
There's also some apparent type duplication between Windows.Win32.Graphics.Dxgi and Windows.Win32.Graphics.Dxgi.Common ... it seems that it's generating a lot of the DXGI stuff _twice_, which I guess is caused by Windows headers having some DXGI stuff and DirectX having it, so the metadata and code generator are considering these belonging to separate namespaces or something. Might not be a big deal but for the fact that when it generates an internal type like DXGI_SWAP_CHAIN_DESC it defines it in Dxgi namespace but its members will be declared as belonging to Dxgi.Common, so I get this weird namespace criss-crossing and alias trouble in my generated code:

// Using Directives:
`using winmdroot = global::Windows.Win32;`

// Inside generated structure def for DXGI_SWAP_CHAIN_DESC:
`internal winmdroot.Graphics.Dxgi.Common.DXGI_MODE_DESC BufferDesc;`

Is this my fault? Do I need to do something in NativeMethods .txt/.json files to deal with this, or is this a legitimate issue?

**Reply by AArnott** [reply]
> I was wondering who the issue belongs to when I find some kind of issue with generation and how to determine if it belongs to the Win32 metadata project or CsWin32 ... for example,

When in doubt, file it at CsWin32 and we can transfer the issue if appropriate. CsWin32 owns C# specific things, and there's quite a set of candidate bugs (not doing varargs right, not using SafeHandle when we should, etc). But when you want an enum instead of an int, or a struct is missing a field, that's a metadata issue and[ you can file it there.

> Also, what about auto-documentation issues? I'm sure that belongs here?

Most of the doc content comes from docs.microsoft.com (or rather, the git repo behind it). But that funnels through the win32metadata site into a nuget package which CsWin32 consumes. We can't do much to improve that. We used a hand-crafted scraping tool to get the docs, so it's imperfect and incomplete. 
CsWin32 owns how that data is formatted as C# xml doc comments. And a *very* little bit of docs come straight out of CsWin32. These only apply to the hand-crafted code such as inline arrays, a few helper methods, etc.

**Reply by AArnott** [reply]
> There's also some apparent type duplication 

That would be a metadata issue.

**Reply by atcarter714** [reply]
Roger that, I'll gather my notes and post those soon 📫
